File: Libraries/ART/ARTShape.m
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <React/ARTShape.h>

@implementation ARTShape

- (void)setD:(CGPathRef)d
{
  if (d == _d) {
    return;
  }
  [self invalidate];
  CGPathRelease(_d);
  _d = CGPathRetain(d);
}

- (void)dealloc
{
  CGPathRelease(_d);
}

- (void)renderLayerTo:(CGContextRef)context
{
  if ((!self.fill && !self.stroke) || !self.d) {
    return;
  }

  CGPathDrawingMode mode = kCGPathStroke;
  if (self.fill) {
    if ([self.fill applyFillColor:context]) {
      mode = kCGPathFill;
    } else {
      CGContextSaveGState(context);
      CGContextAddPath(context, self.d);
      CGContextClip(context);
      [self.fill paint:context];
      CGContextRestoreGState(context);
      if (!self.stroke) {
        return;
      }
    }
  }
  if (self.stroke) {
    CGContextSetStrokeColorWithColor(context, self.stroke);
    CGContextSetLineWidth(context, self.strokeWidth);
    CGContextSetLineCap(context, self.strokeCap);
    CGContextSetLineJoin(context, self.strokeJoin);
    ARTCGFloatArray dash = self.strokeDash;
    if (dash.count) {
      CGContextSetLineDash(context, 0, dash.array, dash.count);
    }
    if (mode == kCGPathFill) {
      mode = kCGPathFillStroke;
    }
  }

  CGContextAddPath(context, self.d);
  CGContextDrawPath(context, mode);
}

@end

File: ReactCommon/fabric/components/art/state/ARTShape.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <react/components/art/ARTShape.h>
#include <react/components/art/ARTElement.h>
#include <react/components/art/conversions.h>

namespace facebook {
namespace react {

bool ARTShape::operator==(const ARTElement &rhs) const {
  if (rhs.elementType != ARTElementType::Shape) {
    return false;
  }
  auto shape = (const ARTShape &)(rhs);
  return std::tie(
             elementType,
             opacity,
             transform,
             d,
             stroke,
             strokeDash,
             fill,
             strokeWidth,
             strokeCap,
             strokeJoin) ==
      std::tie(
             shape.elementType,
             shape.opacity,
             shape.transform,
             shape.d,
             shape.stroke,
             shape.strokeDash,
             shape.fill,
             shape.strokeWidth,
             shape.strokeCap,
             shape.strokeJoin);
}

bool ARTShape::operator!=(const ARTElement &rhs) const {
  return !(*this == rhs);
}

#ifdef ANDROID
folly::dynamic ARTShape::getDynamic() const {
  return toDynamic(*this);
}
#endif

} // namespace react
} // namespace facebook
